src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1207 @@
+		return rc;
+	rc = ds_pool_check_leader(agg_param->ap_pool_info.api_pool_uuid,
+				  &entry->ie_oid,
+				  agg_param->ap_pool_info.api_pool_version);
+	if (rc == 1) {
+		if (agg_param->ap_agg_entry == NULL) {
+			D_ALLOC_PTR(agg_param->ap_agg_entry);
+			if (entry == NULL) {
+				rc = -DER_NOMEM;
+				goto out;
+			}
+			D_INIT_LIST_HEAD(&agg_param->ap_agg_entry->
+					 ae_cur_stripe.as_dextents);
+
+		}
+		agg_init_entry(agg_param->ap_cont_handle,
+			       agg_param->ap_pool_info.api_cont_hdl,
+			       agg_param->ap_agg_entry, entry, oca,
+			       &agg_param->ap_sgl);
+		rc = agg_subtree_iterate(ih, &entry->ie_oid, agg_param);
+		if (rc)
+			D_ERROR("Subtred iterate failed "DF_RC"\n", DP_RC(rc));
+	} else if (rc < 0) {
+		D_ERROR("ds_pool_check_leader failed "DF_RC"\n", DP_RC(rc));
+		rc = 0;
+	}
+out:
+	return rc;
+}
+
+/* Captures the IV values need for pool and container open. Runs in
+ * system xstream.
+ */
+static void
+agg_iv_ult(void *arg)
+{
+	struct ec_agg_param	*agg_param = (struct ec_agg_param *)arg;
+	daos_prop_t		*prop = NULL;
+	struct daos_prop_entry	*entry;
+	int			 rc = 0;
+
+	rc = ds_pool_iv_srv_hdl_fetch(agg_param->ap_pool_info.api_pool,
+				      &agg_param->ap_pool_info.api_poh_uuid,
+				      &agg_param->ap_pool_info.api_coh_uuid);
+	if (rc) {
+		D_ERROR("ds_pool_iv_srv_hdl_fetch failed: "DF_RC"\n",
+			DP_RC(rc));
+		goto out;
+	}
+
+	D_ALLOC_PTR(prop);
+	if (prop == NULL) {
+		D_ERROR("Property allocation failed\n");
+		rc = -DER_NOMEM;
+		goto out;
+	}
+
+	rc = ds_pool_iv_prop_fetch(agg_param->ap_pool_info.api_pool, prop);
+	if (rc) {
+		D_ERROR("ds_pool_iv_prop_fetch failed: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	entry = daos_prop_entry_get(prop, DAOS_PROP_PO_SVC_LIST);
+	D_ASSERT(entry != NULL);
+	agg_param->ap_pool_info.api_svc_list =
+					(d_rank_list_t *)entry->dpe_val_ptr;
+
+	ABT_eventual_set(agg_param->ap_pool_info.api_eventual,
+			 (void *)&rc, sizeof(rc));
+out:
+	D_FREE(prop);
+}
+
+/* Iterates entire VOS. Invokes nested iterator to recurse through trees
+ * for all objects meeting the criteria: object is EC, and this target is
+ * leader.
+ */
+static int
+agg_iterate_all(struct ds_cont_child *cont)
+{
+	vos_iter_param_t	 iter_param = { 0 };
+	struct vos_iter_anchors  anchors = { 0 };
+	struct ec_agg_param	 agg_param = { 0 };
+	daos_handle_t		 ph = DAOS_HDL_INVAL;
+	int			*status;
+	int			 rc = 0;
+
+	uuid_copy(agg_param.ap_pool_info.api_pool_uuid,
+		  cont->sc_pool->spc_uuid);
+	uuid_copy(agg_param.ap_pool_info.api_cont_uuid, cont->sc_uuid);
+
+	agg_param.ap_pool_info.api_pool_version =
+		cont->sc_pool->spc_pool->sp_map_version;
+	agg_param.ap_pool_info.api_pool = cont->sc_pool->spc_pool;
+	agg_param.ap_cont_handle	= cont->sc_hdl;
+
+	rc = ABT_eventual_create(sizeof(*status),
+				 &agg_param.ap_pool_info.api_eventual);
+	if (rc != ABT_SUCCESS)
+		return dss_abterr2der(rc);
+	rc = dss_ult_create(agg_iv_ult, &agg_param, DSS_ULT_POOL_SRV, 0, 0,
+			    NULL);
```

Comment:
Right now they're called once per invocation of EC aggregation. The EC aggregation function is to be called from container, so I will have to access the variables in container and save them in a struct that will allow passing the values to the EC aggregation.

I discussed this with WangDi, and he agreed that calling it once each invocation of EC aggregation is fine, because the values are cached locally.

Reply:
Note: this code is no longer in the initial functionality patch.